### PR TITLE
Add Autofill Sync15 Logic

### PR DESCRIPTION
**Pull request by lougeniaC64**
Implementing as much of the sync15 logic as possible without the credit card incoming logic and outgoing logic.  We will need to discuss our merge strategy for at minimum this PR and #3842.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `automation/all_tests.sh` runs to completion and produces no failures
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/autofill/src/sync/mod.rs [hunk 1]
```diff
@@ -250,4 +250,209 @@ pub mod test {
 }
 
 #[cfg(test)]
-mod tests;
+mod tests {
```
Comment (mhammond) [inline components/autofill/src/sync/mod.rs:253]: A while back I moved these tests into `test_mod.rs`, and this adds them back here - which I'd be fine with if you felt it better, except for the fact that it didn't touch `test_mod.rs`, so it looks like this has just duplicated them.

**Review by mhammond — COMMENTED** [review]
This is looking in great shape! 🚀

File: components/autofill/src/db/addresses.rs [hunk 2]
```diff
@@ -186,14 +191,210 @@ pub fn touch(conn: &Connection, guid: &Guid) -> Result<()> {
...
 
+fn reset_in_tx(conn: &Connection, assoc: &EngineSyncAssociation) -> Result<()> {
+    // Remove all synced addresses and pending tombstones, and mark all
+    // local addresses as new.
+    conn.execute_batch(
+        "DELETE FROM addresses_mirror;
+
+        DELETE FROM addresses_tombstones;
+
+        UPDATE addresses_data
```
[the remark this answers is not in the document]
Comment (lougeniaC64) [reply components/autofill/src/db/addresses.rs:204]: I'm writing this explanation here as a gut check. If an incoming record has a local record but no mirror record, the `IncomingState` is `LocalRecordInfo::Modified` or `LocalRecordInfo::Unmodified`. Modified records are uploaded, but unmodified records [are explicitly not](https://github.com/mozilla/application-services/blob/148ce2827a9e6832ad5ab6f8a2cc7f7bd1485e76/components/autofill/src/sync/mod.rs#L102). As such, I'll create a ticket for this.

[the remark this answers is not in the document]
Comment (mhammond) [reply components/autofill/src/db/addresses.rs:204]: Yep, that sounds correct.

**Review by mhammond — APPROVED** [review]
Thanks!

File: components/autofill/src/db/addresses.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lougeniaC64) [reply components/autofill/src/db/addresses.rs:204]: https://github.com/mozilla/application-services/issues/3861

**Pull request merged**